import sys

configfile: 'conf/config.yaml'
reference = config['reference']
variants = config['variants']
genes = config['genes']
outdir = config['outdir']

rule All:
    input:
        f"{outdir}/variants_genes.vcf",
        f"{outdir}/classified_variants.vcf"

rule ClassifyVariants:
    input:
        reference = reference,
        variants = variants,
    output:
        f"{outdir}/classified_variants.vcf"
    # TODO conda: 
    threads: workflow.cores
    shell:
        """
        bin/anvir classify --k 14 \\
        --reference {input.reference} \\
        --variants {input.variants} \\
        --threads {threads} \\
        --outfile {output} 
        """

rule AnnotateGenes:
    input:
        vcf = rules.ClassifyVariants.output,
        genes = genes
    output:
        f"{outdir}/variants_genes.vcf"
    # TODO conda:
    shell:
        """
        bcftools annotate -a {input.genes} \\
        -c 'CHROM,FROM,TO,GENE' \\
        -h <(echo \'##INFO=<ID=GENE,Number=1,Type=String,Description=\"affected gene\">\') \\
        {input.vcf} > {output}
        """

rule GetAminoAcidChanges:
        
